README
This package has the Vyatta configuration/operational templates and scripts for the config-mgmt system.


debian/changelog
vyatta-config-mgmt (1.3-1) unstable; urgency=medium

  * New LTS branch.

 -- Daniil Baturin <daniil@vyos.io>  Mon, 04 Jan 2021 17:17:46 +0200

vyatta-config-mgmt (0.34+vyos2+current2) unstable; urgency=medium

  [ Christian Poessinger }
  * Switch off cURL "URL globbing parser" (T504)

 -- Christian Poessinger <christian@poessinger.com>  Fri, 22 Dec 2017 18:03:57 +0100

vyatta-config-mgmt (0.34+vyos2+current1) unstable; urgency=medium

  [ Thomas Jepp ]
  * Fix build depends.

  [ Kim Hagen ]

 -- Kim Hagen <kim.sidney@gmail.com>  Sun, 24 Jan 2016 15:06:08 -0500

vyatta-config-mgmt (0.34+vyos2+lithium5) unstable; urgency=low

  [ Elon ]
  * Bug 584 - Allow sftp as copy and commit-archive location destination
  
  [ Alex Harpin ]

 -- Alex Harpin <development@landsofshadow.co.uk>  Mon, 26 Oct 2015 22:43:59 +0000
 
vyatta-config-mgmt (0.34+vyos2+lithium4) unstable; urgency=low

  [ Alex Harpin ]
  * vyatta-config-mgmt: config archive fails on unknown ssh host key
  * vyatta-config-mgmt: set the saved config filename to use vyos

 -- Alex Harpin <development@landsofshadow.co.uk>  Sat, 11 Jul 2015 15:23:22 +0100
 
vyatta-config-mgmt (0.34+vyos2+lithium3) unstable; urgency=low

  * vyatta-config-mgmt: update dh_gencontrol with new development flag

 -- Alex Harpin <development@landsofshadow.co.uk>  Wed, 17 Jun 2015 19:53:19 +0100
 
vyatta-config-mgmt (0.34+vyos2+lithium2) unstable; urgency=low

  * Update maintainer address

 -- Alex Harpin <development@landsofshadow.co.uk>  Fri, 26 Dec 2014 09:34:21 +0000
 
vyatta-config-mgmt (0.34+vyos2+lithium1) unstable; urgency=low

  * New branch

 -- Alex Harpin <development@landsofshadow.co.uk>  Thu, 11 Dec 2014 11:43:51 +0000
 
vyatta-config-mgmt (0.34+vyos1+helium1) unstable; urgency=low

  * New branch

 -- Daniil Baturin <daniil@baturin.org>  Sat, 15 Feb 2014 16:58:31 +0100

vyatta-config-mgmt (0.34+hydrogen1) unstable; urgency=low

  * New branch

 -- Daniil Baturin <daniil@baturin.org>  Sun, 17 Nov 2013 00:19:54 +0100

vyatta-config-mgmt (0.34+daisy1) unstable; urgency=low

  * create daisy branch

 -- John Southworth <john.southworth@vyatta.com>  Sat, 13 Oct 2012 13:30:51 -0700

vyatta-config-mgmt (0.34) unstable; urgency=low

  * new branch

 -- John Southworth <john.southworth@vyatta.com>  Fri, 12 Oct 2012 19:47:09 -0700

vyatta-config-mgmt (0.33) unstable; urgency=low

  * new branch

 -- Deepti Kulkarni <deepti@vyatta.com>  Sat, 03 Mar 2012 02:25:45 -0800

vyatta-config-mgmt (0.32) unstable; urgency=low

  * new branch

 -- Deepti Kulkarni <deepti@vyatta.com>  Thu, 07 Jul 2011 20:56:22 -0700

vyatta-config-mgmt (0.31) unstable; urgency=low

  * simplify commit-archive implementation

 -- An-Cheng Huang <ancheng@vyatta.com>  Tue, 24 May 2011 15:35:42 -0700

vyatta-config-mgmt (0.30) unstable; urgency=low

  * fix showConfig usage for "compare"

 -- An-Cheng Huang <ancheng@vyatta.com>  Mon, 23 May 2011 16:48:19 -0700

vyatta-config-mgmt (0.29) unstable; urgency=low

  * eliminate hard-coded commit hook directory

 -- An-Cheng Huang <ancheng@vyatta.com>  Mon, 16 May 2011 20:18:31 -0700

vyatta-config-mgmt (0.28) unstable; urgency=low

  * Change the compare script to use the new cli-shell-api showConfig
    api

 -- John Southworth <john.southworth@vyatta.com>  Mon, 28 Mar 2011 08:14:29 -0700

vyatta-config-mgmt (0.27) unstable; urgency=low

  * Change to use euid to get login name.
  * Fix 6674: Internal errors for "compare <> <>".

 -- Stig Thormodsrud <stig@vyatta.com>  Wed, 12 Jan 2011 12:19:51 -0800

vyatta-config-mgmt (0.26) unstable; urgency=low

  * new branch

 -- An-Cheng Huang <ancheng@vyatta.com>  Tue, 28 Dec 2010 13:48:25 -0800

vyatta-config-mgmt (0.25) unstable; urgency=low

  * Rename 'remote-archive commit-uri' to 'commit-archive location' per
    design spec review comments.

 -- Stig Thormodsrud <stig@vyatta.com>  Wed, 22 Dec 2010 11:52:49 -0800

vyatta-config-mgmt (0.24) unstable; urgency=low

  * Set COMMIT_VIA for initial commit-revisions setup.

 -- Stig Thormodsrud <stig@vyatta.com>  Sat, 04 Dec 2010 13:40:52 -0800

vyatta-config-mgmt (0.23) unstable; urgency=low

  * Add cfg-version for config-management.

 -- Stig Thormodsrud <stig@vyatta.com>  Fri, 03 Dec 2010 17:52:37 -0800

vyatta-config-mgmt (0.22) unstable; urgency=low

  * Change 'config-mgmt' to 'config-management' to appease support.
  * Return -1 if commit log is missing.

 -- Stig Thormodsrud <stig@vyatta.com>  Fri, 03 Dec 2010 16:53:14 -0800

vyatta-config-mgmt (0.21) unstable; urgency=low

  * Fix 6511: compare perl errors after failed commit.

 -- Stig Thormodsrud <stig@vyatta.com>  Fri, 03 Dec 2010 14:45:32 -0800

vyatta-config-mgmt (0.20) unstable; urgency=low

  * Add newline after diff.

 -- Stig Thormodsrud <stig@vyatta.com>  Wed, 01 Dec 2010 12:15:49 -0800

vyatta-config-mgmt (0.19) unstable; urgency=low

  * Change commit revisions script to use COMMIT_VIA env var.

 -- Stig Thormodsrud <stig@vyatta.com>  Tue, 23 Nov 2010 15:41:41 -0800

vyatta-config-mgmt (0.18) unstable; urgency=low

  * Change commit-confirm to rollback to last saved config rather than
    active.

 -- Stig Thormodsrud <stig@vyatta.com>  Mon, 22 Nov 2010 19:41:14 -0800

vyatta-config-mgmt (0.17) unstable; urgency=low

  * Add number to /etc/commit scripts to set execution ordering.

 -- Stig Thormodsrud <stig@vyatta.com>  Sat, 20 Nov 2010 14:39:35 -0800

vyatta-config-mgmt (0.16) unstable; urgency=low

  * Revert usage of libwww-curl-perl since it doesn't support scp.

 -- Stig Thormodsrud <stig@vyatta.com>  Sat, 20 Nov 2010 14:18:53 -0800

vyatta-config-mgmt (0.15) unstable; urgency=low

  * Filter suprious temp file lines from diff output.


lib/Vyatta/ConfigMgmt.pm
#
# Module: ConfigMgmt.pm
# 
# **** License ****
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as
# published by the Free Software Foundation.
# 
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# This code was originally developed by Vyatta, Inc.
# Portions created by Vyatta are Copyright (C) 2007-2009 Vyatta, Inc.
# All Rights Reserved.
# 
# Author: Stig Thormodsrud
# Date: October 2010
# Description: Common config-mgmt functions
# 
# **** End License ****
#
package Vyatta::ConfigMgmt;

use strict;
use warnings;

our @EXPORT = qw(cm_commit_add_log cm_commit_get_log cm_get_archive_dir 
                 cm_get_lr_conf_file cm_get_lr_state_file 
                 cm_get_commit_hook_dir cm_write_file cm_read_file
                 cm_commit_get_file cm_commit_get_file_name
                 cm_get_max_revs cm_get_num_revs cm_get_last_commit_file 
                 cm_get_last_push_file cm_get_boot_config_file
                 cm_get_config_rb cm_get_config_dir
                );
use base qw(Exporter);

use Vyatta::Config;
use Vyatta::Interface;
use Vyatta::Misc;
use POSIX;
use IO::Zlib;


my $commit_hook_dir  = `cli-shell-api getPostCommitHookDir`;
my $config_dir       = '/opt/vyatta/etc/config';
my $archive_dir      = "$config_dir/archive";
my $config_file      = "$config_dir/config.boot";
my $lr_conf_file     = "$archive_dir/lr.conf";
my $lr_state_file    = "$archive_dir/lr.state";
my $commit_log_file  = "$archive_dir/commits";
my $last_commit_file = "$archive_dir/config.boot";
my $last_push_file   = "$archive_dir/config.boot-push";
my $config_file_rb   = "$archive_dir/config.boot-rollback";

sub cm_get_boot_config_file {
    return $config_file;
}

sub cm_get_config_rb {
    return $config_file_rb;
}

sub cm_get_commit_hook_dir {
    return "$commit_hook_dir/";
}

sub cm_get_archive_dir {
    return $archive_dir;
}

sub cm_get_config_dir {
    return $config_dir;
}

sub cm_get_lr_conf_file {
    return $lr_conf_file;
}

sub cm_get_lr_state_file {
    return $lr_state_file;
}

sub cm_get_last_commit_file {
    return $last_commit_file;
}

sub cm_get_last_push_file {
    return $last_push_file;
}

sub cm_read_file {
    my ($file) = @_;
    my @lines;
    if ( -e $file) {
	open(my $FILE, '<', $file) or die "Error: read $!";
	@lines = <$FILE>;
	close($FILE);
	chomp @lines;
    }
    return @lines;
}

sub cm_write_file {
    my ($file, $data) = @_;

    open(my $fh, '>', $file) || die "Couldn't open $file - $!";
    print $fh $data;
    close $fh;
    return 1;
}

sub cm_get_max_revs {
    my $config = new Vyatta::Config;
    $config->setLevel('system config-management');
    my $revs = $config->returnOrigValue('commit-revisions');
    return $revs;
}

sub cm_get_num_revs {
    return -1 if ! -e $commit_log_file;
    my @lines = cm_read_file($commit_log_file);
    my $num_revs = scalar(@lines);
    $num_revs-- if $num_revs > 0;  # rev files start at 0
    return $num_revs;
}

sub cm_commit_add_log {
    my ($user, $via, $comment) = @_;

    my $time = time();
    if ($comment =~ /\|/) {
        $comment =~ s/\|/\%\%/g;
    }
    my $new_line = "|$time|$user|$via|$comment|";
    my @lines = cm_read_file($commit_log_file);

    my $revs = cm_get_max_revs();
    unshift(@lines, $new_line);   # head push()
    if (defined $revs and scalar(@lines) > $revs) {
        $#lines = $revs-1;
    }
    my $log = join("\n", @lines);
    cm_write_file($commit_log_file, $log);
}

sub cm_commit_get_log {
    my ($brief) = @_;
    
    my @lines = cm_read_file($commit_log_file);
    
    my @commit_log = ();
    my $count = 0;
    foreach my $line (@lines) {
        if ($line !~ /^\|(.*)\|$/) {
            print "Invalid log format [$line]\n";
            next;
        }
        $line = $1;
        my ($time, $user, $via, $comment) = split(/\|/, $line);
        $comment =~ s/\%\%/\|/g;
        if (defined $brief) {
            my $time_str = strftime("%Y-%m-%d_%H:%M:%S", localtime($time));
            $comment = '' if ! defined $comment;
            my $new_line = sprintf("%s %s by %s", $time_str, $user, $via);
            push @commit_log, $new_line;
        } else {
            my $time_str = strftime("%Y-%m-%d %H:%M:%S", localtime($time));
            my $new_line = sprintf("%-2s  %s by %s via %s\n", 
                                   $count, $time_str, $user, $via);
            push @commit_log, $new_line;
            if (defined $comment and $comment ne '' and $comment ne 'commit') {
                push @commit_log, "    $comment\n" 
            }
        }
        $count++;
    }
    return @commit_log;
}

sub cm_commit_get_file_name {
    my ($revnum) = @_;

    my $filename = $archive_dir . "/config.boot." . $revnum . ".gz";
    return $filename;
}

sub cm_commit_get_file {
    my ($revnum) = @_;

    my $max_revs = cm_get_max_revs();
    if (defined $max_revs and $revnum > $max_revs) {
        print "Error: Invalid config revision number\n";
        exit 1;
    }

    my $filename = cm_commit_get_file_name($revnum);
    die "File [$filename] not found." if ! -e $filename;

    my $fh = new IO::Zlib;
    $fh->open($filename, "rb") or die "Error: read $!";
    my @lines = <$fh>;
    $fh->close;
    return join('', @lines);
}

1;


scripts/vyatta-commit-push.pl
#!/usr/bin/perl
#
# **** License ****
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# A copy of the GNU General Public License is available as
# `/usr/share/common-licenses/GPL' in the Debian GNU/Linux distribution
# or on the World Wide Web at `http://www.gnu.org/copyleft/gpl.html'.
# You can also obtain it by writing to the Free Software Foundation,
# Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
# MA 02110-1301, USA.
#
# This code was originally developed by Vyatta, Inc.
# Portions created by Vyatta are Copyright (C) 2010 Vyatta, Inc.
# All Rights Reserved.
#
# Author: Stig Thormodsrud
# Date: October 2010
# Description: Script to push cofig.boot to one or more URIs
#
# **** End License ****
#

use strict;
use warnings;
use lib '/opt/vyatta/share/perl5/';

use Vyatta::Config;
use Vyatta::ConfigMgmt;
use POSIX;
use File::Compare;
use File::Copy;
use URI;
use Sys::Hostname;
use IO::Prompt;


my $debug = 0;

my $config = new Vyatta::Config;
$config->setLevel('system config-management commit-archive');
my @uris = $config->returnOrigValues('location');

if (scalar(@uris) < 1) {
    print "No URI's configured\n";
    exit 0;
}

my $last_push_file = cm_get_last_push_file();
my $tmp_push_file  = "/tmp/config.boot.$$";

my $cmd = 'cli-shell-api showConfig --show-ignore-edit --show-active-only';
system("$cmd > $tmp_push_file");

if (-e $last_push_file and compare($last_push_file, $tmp_push_file) == 0) {
    exit 0;
}

my $timestamp = strftime(".%Y%m%d_%H%M%S", localtime);
my $hostname = hostname();
$hostname = 'vyos' if ! defined $hostname;
my $save_file = "config.boot-$hostname" . $timestamp;

my $source_addr = $config->returnEffectiveValue('source-address');
my $src_opt = "";
if( defined($source_addr) ) {
    $src_opt = "--interface $source_addr";
    print("Using source address $source_addr\n");
}

print "Archiving config...\n";
foreach my $uri (@uris) {
    my $u      = URI->new($uri);
    my $scheme = $u->scheme();
    my $auth   = $u->authority();
    my $path   = $u->path();
    my ($host, $remote, $cmd) = ('', '', '');
    if (defined $auth and $auth =~ /.*\@(.*)/) {
        $host = $1;
    } else {
        $host = $auth if defined $auth;
    }
    $remote .= "$scheme://$host";
    $remote .= "$path" if defined $path;
    print "  $remote ";

    my $rc = 0;
    if ($scheme =~ /^(scp|sftp)$/ ){
        $cmd = "curl -g -s -S -T $tmp_push_file $uri/$save_file";
        $rc = system($cmd);
        if( $rc >> 8 == 51 ){
            my $rsa_key = `ssh-keyscan -t rsa $host 2>/dev/null`;
            print "The authenticity of host '$host' can't be established.\n";
            my $fingerprint = `ssh-keygen -lf /dev/stdin <<< \"$rsa_key\" | awk {' print \$2 '}`;
            chomp $fingerprint;
            print "RSA key fingerprint is $fingerprint.\n";
            if (prompt("Are you sure you want to continue connecting (yes/no) [Yes]? ", -tynd=>"y")) {
                mkdir "$ENV{HOME}/.ssh/",0700 unless -d "$ENV{HOME}/.ssh";
                open(my $known_hosts, ">>", "$ENV{HOME}/.ssh/known_hosts")
                    or die "Cannot open known_hosts: $!";
                print $known_hosts "$rsa_key\n";
                close($known_hosts);
                $cmd = "curl $src_opt -g -s -S -T $tmp_push_file $uri/$save_file";
                $rc = system($cmd);
                print "\n";
            }
        }
    } else {
        $cmd = "curl $src_opt -s -T $tmp_push_file $uri/$save_file";
        $rc = system($cmd);
    }

    print "cmd [$cmd]\n" if $debug;
    if ($rc eq 0) {
        print " OK\n";
    } else {
        print " Failed!\n";
    }
}

move($tmp_push_file, $last_push_file);

exit 0;


scripts/vyatta-config-mgmt.pl
#!/usr/bin/perl
#
# **** License ****
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# A copy of the GNU General Public License is available as
# `/usr/share/common-licenses/GPL' in the Debian GNU/Linux distribution
# or on the World Wide Web at `http://www.gnu.org/copyleft/gpl.html'.
# You can also obtain it by writing to the Free Software Foundation,
# Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
# MA 02110-1301, USA.
#
# This code was originally developed by Vyatta, Inc.
# Portions created by Vyatta are Copyright (C) 2010 Vyatta, Inc.
# All Rights Reserved.
#
# Author: Stig Thormodsrud
# Date: October 2010
# Description: Script for config management features.
#
# **** End License ****
#

use strict;
use warnings;
use lib '/opt/vyatta/share/perl5/';

use Vyatta::Config;
use Vyatta::ConfigMgmt;
use Getopt::Long;
use File::Basename;
use File::Copy;
use URI;
use IO::Prompt;
use Sys::Syslog qw(:standard :macros);

my $commit_uri_script  = '/opt/vyatta/sbin/vyatta-commit-push.pl';
my $commit_revs_script = '/opt/vyatta/sbin/vyatta-commit-revs.pl';

my $commit_hook_dir  = cm_get_commit_hook_dir();
my $archive_dir      = cm_get_archive_dir();
my $config_file      = "$archive_dir/config.boot";
my $lr_conf_file     = cm_get_lr_conf_file();
my $confirm_job_file = '/var/run/confirm.job';

my $debug = 0;

sub get_link {
    my ($path) = @_;

    my $script = basename($path);
    if ($script =~ /revs/) {
        $script = "01" . $script;
    } elsif ($script =~ /push/) {
        $script = "02" . $script;
    }
    my $link = $commit_hook_dir . $script; 
    return $link;
}

sub check_integer {
    my ($num) = @_;

    if ($num !~ /^\d+$/) {
        print "Invalid number [$num]\n";
        exit 1;
    }
    return 1;
}

sub check_valid_rev {
    my ($rev) = @_;

    check_integer($rev);
    my $num_revs = cm_get_num_revs();
    return 1 if $rev <= $num_revs;
    print "Invalid revision [$rev]\n";
    exit 1;
}

sub parse_at_output {
    my @lines = @_;
    foreach my $line (@lines) {
        if ($line =~ /error/) {
            return (1, '', '');
        } elsif ($line =~ /job (\d+) (.*)$/) {
            return (0, $1, $2);
        } 
    }
    return (1, '', '');
}

sub filter_file_lines {
    my ($diff) = @_;

    my @lines = split("\n", $diff);
    my $line1 = shift @lines;
    my $line2 = shift @lines;
    unshift(@lines, $line2) if $line2 !~ /^\+\+\+ /;
    unshift(@lines, $line1) if $line1 !~ /^\-\-\- /;
    return join("\n", @lines);
}

sub filter_version_string {
    my ($diff) = @_;

    # find last diff hunk, skip it if it's the version string
    my @lines = split("\n", $diff);
    my @last_hunk = ();
    my $found = 0;
    while (my $line = pop(@lines)) {
        unshift(@last_hunk, $line);
        $found = 1 if $line =~ /vyatta-config-version/;
        last if $line =~ /^@@/;
    }
    return join("\n", @lines) if $found;
    push(@lines, @last_hunk);
    return join("\n", @lines);
}


#
# main
#
my ($action, $uri, $revs, $revnum, $minutes, $filename);

Getopt::Long::Configure('pass_through');
GetOptions("action=s"      => \$action,
           "uri=s"         => \$uri,
           "revs=s"        => \$revs,
           "revnum=s"      => \$revnum,
           "minutes=s"     => \$minutes,
           "file=s"        => \$filename,
);

die "Error: no action"      if ! defined $action;

my ($cmd, $rc) = ('', 1);

if ($action eq 'update-uri') {
    print "update-uri\n" if $debug;
    my $config = new Vyatta::Config;
    $config->setLevel('system config-management commit-archive');
    my @uris = $config->returnValues('location');
    my $link = get_link($commit_uri_script);
    if (scalar(@uris) > 0 and ! -e $link) {
        print "add link [$link]\n" if $debug;
        $rc = system("ln -s $commit_uri_script $link");
        exit $rc;
    } elsif (scalar(@uris) < 1 and -e $link) {
        print "remove link [$link]\n" if $debug;
        $rc = system("rm -f $link");
        exit $rc;
    }
    exit 0;
}

if ($action eq 'valid-uri') {
    die "Error: no uri" if ! defined $uri;
    print "valid-uri [$uri]\n" if $debug;
    my $u = URI->new($uri);
    exit 1 if ! defined $u;
    my $scheme = $u->scheme();
    my $auth   = $u->authority();
    my $path   = $u->path();
    
    exit 1 if ! defined $scheme or ! defined $path;
    if ($scheme eq 'tftp') {
    } elsif ($scheme eq 'ftp') {
    } elsif ($scheme eq 'scp') {
    } elsif ($scheme eq 'sftp') {
    } else {
        print "Unsupported URI scheme\n";
        exit 1;
    }
    exit 0;
}

if ($action eq 'update-revs') {
    die "Error: no revs" if ! defined $revs;
    print "update-revs [$revs]\n" if $debug;
    check_integer($revs);
    my $link = get_link($commit_revs_script);
    if ($revs == 0) {
        print "remove link [$link]\n" if $debug;
        $rc = system("rm -f $link");
    } else {
        if (! -e $link) {
            print "add link [$link]\n" if $debug;
            $rc = system("ln -s $commit_revs_script $link");
        }
        if (! -e $archive_dir) {
            system("sudo mkdir $archive_dir");
            system("sudo chgrp vyattacfg $archive_dir");
            system("sudo chmod 775 $archive_dir");
        }
        my $lr_conf = "$config_file {\n";
        $lr_conf   .= "\t su root vyattacfg\n";
        $lr_conf   .= "\t rotate $revs\n";
        $lr_conf   .= "\t start 0\n";
        $lr_conf   .= "\t compress \n";
        $lr_conf   .= "\t copy \n";
        $lr_conf   .= "}\n";
        cm_write_file($lr_conf_file, $lr_conf);

        # logrotate doesn't like when its config is writeable
        # for anyone but root
        system("sudo chmod 644 $lr_conf_file");

        my $num_revs = cm_get_num_revs();
        if (! -e "$archive_dir/commits" or $num_revs == 0) {
            # store a baseline config
            system("sudo touch $archive_dir/commits");
            system("sudo chgrp vyattacfg $archive_dir/commits");
            system("sudo chmod 664 $archive_dir/commits");
            my $cmd = "$commit_revs_script baseline config.boot";
            system("sudo sg vyattacfg \"export COMMIT_VIA=init; $cmd\"");
        }
        exit 0;
    }
    exit 0;
}

if ($action eq 'show-commit-log') {
    print "show-commit-log\n" if $debug;
    my $max_revs = cm_get_max_revs();
    if (!defined $max_revs or $max_revs <= 0) {
        print "commit-revisions is not configured.\n\n";
    }
    my @log = cm_commit_get_log();
    foreach my $line (@log) {
        print $line;
    }
    exit 0;
}

if ($action eq 'show-commit-log-brief') {
    print "show-commit-log-brief\n" if $debug;
    my $max_revs = cm_get_max_revs();
    my @log = cm_commit_get_log(1);
    foreach my $line (@log) {
        $line =~ s/\s/_/g;
        print $line, ' ';
    }
    exit 0;
}

if ($action eq 'show-commit-file') {
    die "Error: no revnum" if ! defined $revnum;
    print "show-commit-file [$revnum]\n" if $debug;
    check_valid_rev($revnum);
    my $file = cm_commit_get_file($revnum);
    print $file;
    exit 0;
}

if ($action eq 'diff') {
    print "diff\n" if $debug;    
    my $args = $#ARGV;
    if ($args < 0) {
        my $rc = system("cli-shell-api sessionChanged");
        if (defined $rc and $rc > 0) {
            print "No changes between working and active configurations\n";
            exit 0;
        }
        my $show_args = '--show-show-defaults --show-context-diff';
        # default behavior for showConfig is @ACTIVE vs. @WORKING, so no
        # need to write to a file first
        my $diff = `cli-shell-api showConfig $show_args`;
        if (defined $diff and length($diff) > 0) {
            print "$diff";
        } else {
            print "No changes between working and active configurations\n";
            exit 0;
        }
    } elsif ($args eq 0) {
        my $rev1 = $ARGV[0];
        check_valid_rev($rev1);
        my $filename1 = cm_commit_get_file_name($rev1);
        my $outfile = $filename1;
        $outfile =~ s/(.*)\.gz/$1/g;
        system("zcat $filename1 > $outfile");
        my $diff = `cli-shell-api showConfig --show-cfg1 $outfile --show-cfg2 \@WORKING --show-show-defaults --show-context-diff`;
        if (defined $diff and length($diff) > 0) {
            print "$diff";
        } else {
            print "No changes between working and "
                . "revision $rev1 configurations\n";
        }
        system("rm $outfile");
    } elsif ($args eq 1) {
        my $rev1 = $ARGV[0];
        my $rev2 = $ARGV[1];
        check_valid_rev($rev1);
        check_valid_rev($rev2);
        my $filename  = cm_commit_get_file_name($rev1);
        my $filename2 = cm_commit_get_file_name($rev2);
        my $outfile = $filename;
        $outfile =~ s/(.*)\.gz/$1/g;
        system("zcat $filename > $outfile");
        my $outfile2 = $filename2;
        $outfile2 =~ s/(.*)\.gz/$1/g;
        system("zcat $filename2 > $outfile2");
        my $diff = `cli-shell-api showConfig --show-cfg1 $outfile2 --show-cfg2 $outfile --show-show-defaults --show-context-diff`;
        if (defined $diff and length($diff) > 0) {
            print "$diff";
        } else {
            print "No changes between revision $rev1 and "
                . "revision $rev2 configurations\n";
        }
        system("rm $outfile2");
        system("rm $outfile");
    } elsif ($args eq 2) {
        my $rev1 = $ARGV[0];
        my $rev2 = $ARGV[1];
        check_valid_rev($rev1);
        check_valid_rev($rev2);
        my $filename  = cm_commit_get_file_name($rev1);
        my $filename2 = cm_commit_get_file_name($rev2);
        my $outfile = $filename;
        $outfile =~ s/(.*)\.gz/$1/g;
        system("zcat $filename > $outfile");
        my $outfile2 = $filename2;
        $outfile2 =~ s/(.*)\.gz/$1/g;
        system("zcat $filename2 > $outfile2");
        my $diff = `cli-shell-api showConfig --show-cfg1 $outfile2 --show-cfg2 $outfile --show-commands --show-show-defaults --show-context-diff`;
        if (defined $diff and length($diff) > 0) {
            my @difflines = split('\n', $diff);
            foreach my $line (@difflines){
              my @words = split(' ', $line);
              my $elements = scalar(@words);
              my @non_leaf = @words[0 .. ($elements - 2)] ;
              my $path = join(' ', @non_leaf);
              $path =~ s/'//g;
              my $cmd = "$path " . @words[($elements - 1)];
              print "$cmd\n";
            }
        } else {
            print "No changes between revision $rev1 and "
                . "revision $rev2 configurations\n";
        }
        system("rm $outfile2");
        system("rm $outfile");
    }
    exit 0;
}

if ($action eq 'commit-confirm') {
    die "Error: no minutes" if ! defined $minutes;
    print "commit-confirm [$minutes]\n" if $debug;
    if (-e $confirm_job_file) {
        print "Another confirm is pending\n";
        exit 0;
    }
    my $max_revs = cm_get_max_revs();
    if (!defined $max_revs or $max_revs <= 0) {
        print "commit-revisions is not configured.\n\n";
        exit 1;
    }
    check_integer($minutes);
    print "commit confirm will be automatically reboot in $minutes"
        . " minutes unless confirmed\n";
    if (prompt("Proceed? [confirm]", -y1d=>"y")) {
    } else {
        print "commit-confirm canceled\n";
        exit 1;
    }

    my @lines= cm_read_file(cm_get_boot_config_file());
    my $rollback_config  = join("\n", @lines);
    $rollback_config .= "\n";
    my $config_rb = cm_get_config_rb();
    cm_write_file($config_rb, $rollback_config);
    
    $cmd = "/opt/vyatta/sbin/vyatta-config-mgmt.pl --action rollback" 
         . " --file $config_rb";
    @lines = `echo sudo sg vyattacfg \\"$cmd\\" | at now + $minutes minutes 2>&1`;
    my ($err, $job, $time) = parse_at_output(@lines);
    if ($err) {
        print "Error: unable to schedule reboot\n";
        exit 1;
    }
    system("echo $job > $confirm_job_file");
    exit 0;
}

if ($action eq 'confirm') {
    if (! -e $confirm_job_file) {
        print "No confirm pending\n";
        exit 0;
    }
    my $job = `cat $confirm_job_file`;
    chomp $job;
    system("sudo atrm $job");
    system("sudo rm -f $confirm_job_file");
    # log confirm
    exit 0;
}

if ($action eq 'rollback') {
    my ($method, $rollback_config) = (undef, undef);

    if (defined $revnum) {
        print "rollback [$revnum]\n" if $debug;
        check_valid_rev($revnum);        
        $method = 'revnum';
        if (prompt("Proceed with reboot? [confirm]", -y1d=>"y")) {
        } else {
            print "Cancelling rollback\n";
            exit 0;
        }
        $rollback_config = cm_commit_get_file($revnum);
    }

    if (defined $filename) {
        print "rollback [$filename]\n" if $debug;
        if (! -e $filename) {
            die "Error: file [$filename] doesn't exist";
        }
        if (defined $method) {
            die "Error: can only define revnum or file";
        }
        $method = 'file';
        # Should have code to validate config, but for now only
        # called internally.  If we later expose this to cli
        # we'll need to prompt for confirmation.
        my @lines = cm_read_file($filename);
        $rollback_config  = join("\n", @lines);
        $rollback_config .= "\n";
    }
    if (!defined $method) {
        die "Error: must define either revnum or file";
    }

    my ($user) = getpwuid($<);
    my $boot_config_file = cm_get_boot_config_file();
    my $archive_dir      = cm_get_archive_dir();
    my $last_commit_file = cm_get_last_commit_file();
    system("sudo cp $boot_config_file $archive_dir/config.boot-prerollback");
    cm_write_file($boot_config_file, $rollback_config);
    cm_write_file($last_commit_file, $rollback_config); # white lie
    my $cmd = "$commit_revs_script --rollback=1 rollback/reboot";
    system("sudo sg vyattacfg \"$cmd\"");
    openlog($0, "", LOG_USER);
    my $login = getpwuid($<) || "unknown";
    syslog("warning", "Rollback reboot requested by $login");
    closelog();
    exec("sudo /sbin/reboot");
}

exit $rc;

# end of file
